### PR TITLE
add custom `User-Agent` header to `Config`

### PR DESCRIPTION
**Pull request by Air-t**
Allow easy override of the `User-Agent` header with the `Config` class.

### CODE REVIEW

**Review by IKarbowiak — DISMISSED** [review]

**Review by maarcingebala — DISMISSED** [review]

File: requests_hardened/client.py [hunk 1]
```diff
@@ -30,9 +30,10 @@ def send(self, request: PreparedRequest, **kwargs: Any) -> Response:
 
     def prepare_request(self, request: Request) -> PreparedRequest:
         url = request.url
-
         if self._config.ip_filter_enable is True:
             headers = request.headers or {}
+            if self._config.user_agent_override:
```
Comment (NyanKiyoshi) [inline requests_hardened/client.py:35]: `None` should probably be the only case where the override is disabled. An empty user-agent is valid and will be accepted by `requests` library

```suggestion
            if self._config.user_agent_override is not None:
```

File: requests_hardened/config.py [hunk 2]
```diff
@@ -18,3 +18,6 @@ class Config:
 
     # The default timeout value to set to all requests.
     default_timeout: Optional[T_TIMEOUT_TUPLE]
+
+    # override default User-Agent header
```
Comment (NyanKiyoshi) [inline requests_hardened/config.py:22]:
```suggestion
    # Override the default User-Agent header of the requests library.
```

**Review by NyanKiyoshi — CHANGES_REQUESTED** [review]
Add test cases and update documentation (README.rst)

**Review by maarcingebala — DISMISSED** [review]

File: requests_hardened/client.py [hunk 3]
```diff
@@ -30,10 +30,8 @@ def send(self, request: PreparedRequest, **kwargs: Any) -> Response:
...
     def prepare_request(self, request: Request) -> PreparedRequest:
         url = request.url
-
+        headers = request.headers or {}
         if self._config.ip_filter_enable is True:
-            headers = request.headers or {}
-
             # Cast potentially immutable header list to `dict`
             if not isinstance(headers, dict):
                 headers = cast(dict, dict(**headers))
```
Comment (NyanKiyoshi) [inline requests_hardened/client.py:37]: This should be done after `headers = request.headers or {}`

File: requests_hardened/client.py [hunk 4]
```diff
@@ -49,4 +47,8 @@ def prepare_request(self, request: Request) -> PreparedRequest:
             )
             request.url = url
             request.headers = headers
-        return super().prepare_request(request)
+        prep = super().prepare_request(request)
+        if self._config.user_agent_override is not None:
+            prep.headers.update({"User-Agent": self._config.user_agent_override})
```
Comment (NyanKiyoshi) [inline requests_hardened/client.py:52]: Is it really needed to change headers after `super().prepare_request(request)`? If not, I suggest just changing `request.headers`

Comment (Air-t) [reply requests_hardened/client.py:52]: Unfortunately, this will fail to set proper headers - that was a good point to add the test cases.

Comment (NyanKiyoshi) [reply requests_hardened/client.py:52]: Can you add a comment explaining why in the code then?

Comment (Air-t) [reply requests_hardened/client.py]: Ok, found the issue - it was the matter of which headers needed to be updated

File: requests_hardened/client.py [hunk 5]
```diff
@@ -30,14 +30,12 @@ def send(self, request: PreparedRequest, **kwargs: Any) -> Response:
 
     def prepare_request(self, request: Request) -> PreparedRequest:
         url = request.url
+        headers = request.headers or {}
+        # Cast potentially immutable header list to `dict`
+        if not isinstance(headers, dict):
+            headers = cast(dict, dict(**headers))
```
Comment (NyanKiyoshi) [inline requests_hardened/client.py:36]: Now we can move it back under the block `if self._config.ip_filter_enable is True:` 😅

**Review by NyanKiyoshi — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

File: README.rst [hunk 6]
```diff
@@ -28,7 +28,7 @@ that can have a security impact:
 
 - ``Config.never_redirect = False`` always reject HTTP redirects
 - ``Config.default_timeout = (2, 10)`` sets the default timeout value when no value or ``None`` is passed
-
+- ``Config.user_agent_override = None`` optional config to override ``User-Agent`` header. When set to ``None``, `requests` library will set its `default user-agent <https://github.com/psf/requests/blob/ee93fac6b2f715151f1aa9a1a06ddba9f7dcc59a/src/requests/utils.py#L886-L892>`_.
```
Comment (NyanKiyoshi) [inline README.rst:31]: Slight issue:

```suggestion
- ``Config.user_agent_override = None`` optional config to override ``User-Agent`` header. When set to ``None``, ``requests`` library will set its `default user-agent <https://github.com/psf/requests/blob/ee93fac6b2f715151f1aa9a1a06ddba9f7dcc59a/src/requests/utils.py#L886-L892>`_.
```

**Review by maarcingebala — APPROVED** [review]

**Pull request closed without merging**